tests/test_mig_shared_base.py

```diff
@@ -0,0 +1,92 @@
+# -*- coding: utf-8 -*-
+#
+# --- BEGIN_HEADER ---
+#
+# test_mig_shared_compat - unit test of the corresponding mig shared module
+# Copyright (C) 2003-2024  The MiG Project by the Science HPC Center at UCPH
+#
+# This file is part of MiG.
+#
+# MiG is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# MiG is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
+# USA.
+#
+# --- END_HEADER ---
+#
+
+"""Unit tests for the migrid module pointed to in the filename"""
+
+import binascii
+import os
+import sys
+
+from tests.support import PY2, MigTestCase, testmain
+
+from mig.shared.base import force_default_fs_coding_rec, force_default_str_coding_rec
+
+DUMMY_BYTECHARS = b'DEADBEEF'
+DUMMY_BYTESRAW = binascii.unhexlify('DEADBEEF') # 4 bytes
+DUMMY_UNICODE = u'UniCode123½¾µßðþđŋħĸþł@ª€£$¥©®'
+
+
+class MigSharedBase__force_default_fs_coding_rec(MigTestCase):
+    """Unit tests of mig.shared.base force_default_fs_coding_rec()"""
+
+    def test_encode_a_string(self):
+        output = force_default_fs_coding_rec('foobar')
+
+        self.assertEqual(output, 'foobar')
```

Comment:
I believe this testing is insufficient and masks the point raised above. The `default_fs_coding` happens to always be utf8, so on py3 `output` should definitely not be the (implicitly unicode) string `'foobar'`.

I'd say we need to test with both explicit byte and unicode input **and** compare the output to that of the simple non-recursive `force_default_fs_coding`. 
As an example I think we could use a general template like:
```
    def test_encode_a_string(self):
	input_vals = (b'foobar', u'foobar')
        truth = force_default_fs_coding('foobar')
        for val in input_vals:
            output = force_default_fs_coding_rec(val)
            self.assertEqual(output, truth)

    def test_encode_within_a_dict(self):
        input_vals = ({b'key': b'value'}, {u'key': u'value'})
        truth = {force_default_fs_coding('key'):
                 force_default_fs_coding('value')}
        for val in input_vals:
            output = force_default_fs_coding_rec(val)
            self.assertEqual(output, truth)
```
for each of the test methods and only replace the `input_vals` and `truth` to the actual types tested.

If we want to be sure to stand on more _solid_ ground by not only comparing to a generated `truth`, we could e.g. add similar `force_utf8` and `force_unicode` test cases that explicitly compare to `b'bla'` and `u'bla'` ground truths respectively.

Reply:
It took me some pondering but I think I actually figured out the more general thing tat led me astray in this PR - turns out I previously had all of this "right" and then, when I went to reconcile it with Py 2 the string/bytes duality messed me up and I made everything look like strings. I've gone with explicit b'' strings and it made the problem crystal clear.